### DISCUSSION TITLE
GraphQL

### DISCUSSION THREAD

**Discussion by natew** [pocketbase/pocketbase#9] [Q&A]
This would be a really cool feature :)

**Comment by ganigeorgiev** [comment]
Not implementing a GraphQL API was an early decision.

In my opinion, it will add a fair amount of complexity without too much benefit for the end user.
Furthermore, PocketBase comes with its own small search query language that shares the same syntax with the collection's access rules, making the dev experience a little more consistent across the application (and at least for me, easier to read and write).

In any case, the SDKs (currently only [JavaScript](https://github.com/pocketbase/js-sdk)) are intended to simplify the interaction with the Web API and ideally the users shouldn't be concerned whether the API is REST or GraphQL.

**Reply by natew** [reply]
Querying SQL from GraphQL is a big upgrade in frontend development across a number of ways. 

Recommend checking out https://gqty.dev which shows how elegant you can make data fetching. Combined with something like this you’d have an incredible tool for simple intranet type tools or one off apps.

**Reply by ganigeorgiev** [reply]
I'm not against GraphQL in general, just I'm not sure that for this project the added complexity will be worthed (it will also require runtime code generation for the user defined api schemas).

**Reply by sultaniman** [reply]
I also second @ganigeorgiev because imo graphql support is a big burden on maintenance, while providing "convenient" DX to frontent part I think it will just steal resources.

**Comment by rodydavis** [comment]
Could TUQL be used to point to the SQLite file?


https://www.the-guild.dev/graphql/mesh/docs/handlers/tuql

**Reply by ganigeorgiev** [reply]
Using something like TUQL is possible but you'll lose the query capabilities and security rules that PocketBase provides.

I haven't tried it, but I can't think of a reason why it shouldn't work. The PocketBase DB file would be usually located in `./pb_data/data.db`.
Behind the scenes, each created collection in PocketBase is just a regular table with the same name and fields (+ some additional system fields like `id`, `created`, `updated`).

**Comment by gevera** [comment]
I know there are plenty of tools besides Strapi that do the same thing (  expose graphQL endpoint), like Directus, Postgraphile or Nhost. Most of them on Node.js though.  And even though, Pocketbase is intended to be minimalist and have REST API and JS SDK, I do consider graphQL to be minimalist as well. You just get the data that you have request - no more, no less.

**Reply by jedashford** [reply]
"You just get the data that you have request - no more, no less" 👍 -- No need for client modeling and relationships that take in vertical resources and merge/massage/filter the data into something the client needs, on EVERY client. GraphQL has been a nice iteration on client/server interaction.

**Comment by jedashford** [comment]
I'll add my 2 cents. Had an engineer suggest this as an option for our team. As a startup years ago we moved to graphql and it was a positive change for us. Some key points: Domain language can be used across the stack, business logic pulled to server side enabling simplistic queries and super clean clients, we pulled in more engineers to work full stack, and allowed a simplistic multi-service architecture out of the box. I could go on for days with the benefits of a graph api vs REST (CRUD) for both clients and servers as a primary client<>server interaction, but maybe this use case isn't the target group for pocketbase. We have plenty of small services that are one file node services that may only use rest, and that maybe pocketbase would cover the smaller use case.

If the decision early on to not build graphql, it might be reevaluated. Not saying you do it, but depends on the market, where the tech is going, and what the target audience is.

**Comment by trm217** [comment]
Honestly, GraphQL not being available is the only thing, that prevents me from using Pocketbase.
Of course, REST is also fine, but the increased convenience & capabilities of GraphQL are a game-changer. I‘ve yet to see a frontend-dev that wished to transition back to REST.

**Reply by tigawanna** [reply]
Tbf the client SDKs do let you select the fields you want + direct and indirect relations getting you get as close to that experience as possible , GraphQL is sweet but it comes with some major overheads that can be unnecessary

**Reply by trm217** [reply]
The question is: do I enjoy using GraphQL and do I want to work with it? Yes.
So I am not going to diverge from a stack that I feel confident, just to accommodate a framework.
It's the other way around.
Therefore I simply can't use Pocketbase.
If it'd have a GraphQL api on the other hand, I definitely would.

**Reply by jonlambert** [reply]
Also worth noting that from the perspective of a newcomer, the pocketbase client appears to have light-touch TypeScript support at best. (Please correct me if I'm wrong!)

Personally I've no desire to leave types behind. Supporting GraphQL would allow users to take advantage of the very mature client ecosystems that already exist in libraries for most languages already: type generation, normalised caching, subscriptions, fragments, federation etc.

**Reply by tigawanna** [reply]
The underrated beauty of this project is the community, there's a project that adds better types 
https://github.com/david-plugge/typed-pocketbase

I believe PRs or helpful wrappers would be the solution to this issue for now as it is an open source project with no company behind it

**Comment by txbrown** [comment]
Having graphql out of the box would be very useful. In my case I already was using graphql in a custom server setup with `gql-gen` and I added pocketbase as a framework. The only annoying part is mapping the `model.Record` to grapqhl, other than that I think the combo works pretty great.

As I think further on this, I believe that pocketbase should focus on the content making it content management and data modelling as slick as possible with the other added features like auth, hooks, etc and compose it with other golang tools like gql-gen.

**Comment by ganigeorgiev** [comment]
_I'm closing this discussion to avoid spamming further the participants. There are no plans for now to implement GraphQL API. If I change my mind I'll update it but it is very unlikely to happen._

**Comment by kinthaiofficial** [comment]
GraphQL on top of PocketBase is an interesting ask — the core tension is that PocketBase's REST API is already quite ergonomic, and GraphQL adds significant complexity for the server to implement.

That said, there are real use cases where GraphQL wins:

**When GraphQL would help:**
- Complex nested queries in a single round-trip: `user { posts { comments { author { name } } } }` vs 4 REST calls
- Frontend teams that want to specify exactly which fields they need (avoids over-fetching)
- When your client is already GraphQL-native (e.g., using Relay or Apollo)

**When REST (PocketBase's current API) is fine:**
- Simple CRUD on individual collections
- Mobile apps where API simplicity matters more than query flexibility
- Projects with a small data model where REST round-trips aren't a bottleneck

**Practical alternatives without native GraphQL:**
- A thin GraphQL proxy layer in front of PocketBase's REST API (graphql-request + custom resolvers) — adds GraphQL semantics without PocketBase needing to implement it
- PocketBase extensions/hooks that implement a `/graphql` endpoint — the hook system might support this

**The real question:** Is the use case driving this "I need to query nested relations efficiently" or "I'm already using GraphQL everywhere"? The former might be better served by improving PocketBase's expand query syntax; the latter is a genuine need.

If this gets implemented, schema introspection (so existing GraphQL tooling works) should be the first priority.

**Reply by ganigeorgiev** [reply]
Honestly I don't know what is wrong with "developers" nowadays...

Go away please.
